package/resource_monitor/resource_monitor/src/query_process_file_descriptor.c

```diff
@@ -0,0 +1,491 @@
+      if (s < 0 || (size_t)s >= sizeof(file_entry->file_str)) {
+#ifdef DEBUG_QUERY_FD_TAB
+        PK_LOG_ANNO(LOG_ERR, "file descriptor query: failed to copy file string");
+#endif
+        return false;
+      }
+
+
+      file_entry = calloc(1, sizeof(file_entry_t));
+      *file_entry = new_entry;
+
+      bool added = file_table_put(state->file_table, file_str, file_entry);
+
+      if (!added) {
+#ifdef DEBUG_QUERY_FD_TAB
+        PK_LOG_ANNO(LOG_ERR, "file descriptor query:failed add entry for pid '%s'", file_str);
+#endif
+        free(file_entry);
+        return false;
+      }
+    }
+    file_entry->fd_count++;
+  }
+  return true;
+}
+
+static bool parse_fd_line(resq_state_t *state, const char *line)
+{
+  char id_str[16] = {0};
+  char exe_str[32] = {0};
+  char file_str[64] = {0};
+  line_spec_t line_specs[FD_STATE_COUNT] = {[FD_STATE_ID_STR] =
+                                              (line_spec_t){
+                                                .type = FT_STR,
+                                                .dst.str = id_str,
+                                                .buflen = sizeof(id_str),
+                                                .desc = "pid",
+                                                .next = FD_STATE_EXE_FILE_STR,
+                                              },
+                                            [FD_STATE_EXE_FILE_STR] =
+                                              (line_spec_t){
+                                                .type = FT_STR,
+                                                .dst.str = exe_str,
+                                                .buflen = sizeof(exe_str),
+                                                .desc = "executable",
+                                                .next = FD_STATE_FD_STR,
+                                              },
+                                            [FD_STATE_FD_STR] =
+                                              (line_spec_t){
+                                                .type = FT_STR,
+                                                .dst.str = file_str,
+                                                .buflen = sizeof(file_str),
+                                                .desc = "file descriptor",
+                                                .next = FD_STATE_DONE,
+                                              },
+                                            [FD_STATE_DONE] = (line_spec_t){
+                                              .desc = "done",
+                                            }};
+
+  if (!parse_tab_line(line, FD_STATE_ID_STR, FD_STATE_DONE, line_specs, "\t")) {
+#ifdef DEBUG_QUERY_FD_TAB
+    PK_LOG_ANNO(LOG_ERR, "file descriptor query: parse_tab_line failed ");
+#endif
+    return false;
+  }
+  return process_fd_entry(state, file_str, id_str);
+}
+
+static void destroy_table_entry(table_t *table, void **entry)
+{
+  (void)table;
+
+  pid_entry_t *pid_entry = *entry;
```

Comment:
Let's remove this cast since `destory_table_entry` is used by the file table and the pid table.